### PR TITLE
fix: move state into ChainMetadatastore to avoid race condition

### PR DESCRIPTION
**Pull request by kevin-satsuma**
# Description

There is race condition in node_filter that comes from block height states being stored in two separate places. Current block height and block height errors come from the BlockHeightChecker while global/group max height come from ChainMetadataStore.

We can remove the race condition by moving all state into ChainMetadataStore and providing a single atomic get function for all block height related data.

The PR is a little large due to the mock generation.

CC @IvanVergiliev @dianwen 

## Type of change

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] 😎 New feature (non-breaking change which adds functionality)
- [ ] ⁉️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] ⚒️ Refactor (no functional changes)
- [ ] 📖 Documentation (updating or adding docs)

# How Has This Been Tested?

Running locally.

### CODE REVIEW

File: internal/metadata/store.go [hunk 1]
```diff
@@ -27,40 +37,46 @@ func (c *ChainMetadataStore) Start() {
...
+		c.updateHeightForGroup(groupID, blockHeight)
+		c.updateHeightForUpstream(groupID, updateID, blockHeight)
+		c.updateErrorForUpstream(groupID, updateID, nil)
 	}
-
-	return <-returnChannel
 }
 
-func (c *ChainMetadataStore) ProcessUpdate(update BlockHeightUpdate) {
+func (c *ChainMetadataStore) ProcessErrorUpdate(groupID, updateID string, err error) {
```
Comment (brianluong) [inline internal/metadata/store.go:78]: nit: s/updateID/upstreamID?

File: internal/metadata/store.go [hunk 2]
```diff
@@ -27,40 +37,46 @@ func (c *ChainMetadataStore) Start() {
...
+		returnChannel <- blockHeightStatus
 		close(returnChannel)
 	}
 
 	return <-returnChannel
 }
 
-func (c *ChainMetadataStore) GetMaxHeightForGroup(groupID string) uint64 {
-	returnChannel := make(chan uint64)
+func (c *ChainMetadataStore) ProcessBlockHeightUpdate(groupID, updateID string, blockHeight uint64) {
```
Comment (brianluong) [inline internal/metadata/store.go:69]: nit: s/updateID/upstreamID?

File: internal/route/node_filter.go [hunk 3]
```diff
@@ -118,66 +117,57 @@ func (f *IsCloseToGlobalMaxHeight) Apply(
...
 
 		return false
 	}
 
-	maxHeight := f.chainMetadataStore.GetGlobalMaxHeight()
-	upstreamHeight := check.GetBlockHeight()
-	isClose := upstreamHeight+f.maxBlocksBehind >= maxHeight
-
-	if isClose {
+	if status.BlockHeight+f.maxBlocksBehind >= status.GlobalMaxBlockHeight {
```
Comment (brianluong) [inline internal/route/node_filter.go:131]: nit: should we leave the `isClose` variable here or add a comment on this? Mainly thinking about readability. The function name is `IsClose...` so I think it help with tying together the key line(s) of code w/ the function's main purpose.

Comment (brianluong) [inline internal/route/node_filter.go:131]: nit: should we leave the `isClose` variable here or add a comment on this? Mainly thinking about readability. Although adding a var for this is arguably redundant, the function name is `IsClose...` so I think it help with tying together the key line(s) of code w/ the function's main purpose. Up to you tho

File: internal/metadata/store.go [hunk 4]
```diff
@@ -27,40 +37,46 @@ func (c *ChainMetadataStore) Start() {
...
 }
 
-func (c *ChainMetadataStore) GetGlobalMaxHeight() uint64 {
-	returnChannel := make(chan uint64)
+func (c *ChainMetadataStore) GetBlockHeightStatus(groupID, upstreamID string) BlockHeightStatus {
+	returnChannel := make(chan BlockHeightStatus)
 	c.opChannel <- func() {
-		returnChannel <- c.globalMaxHeight
+		blockHeightStatus := BlockHeightStatus{
+			Error:                c.errorByUpstreamID[groupID+upstreamID],
```
Comment (brianluong) [inline internal/metadata/store.go:55]: Good looks on this.

In other parts of the code, we assume that upstreamID is unique, [for example here](https://github.com/satsuma-data/node-gateway/blob/e6e71622712e4563e91a66ff42986b2b00a3891b/internal/checks/manager.go#L82). 

Although currently, we don't fail out if we define duplicate upstreamIDs in the config. So maybe this should be a check when validating the config. We validate for uniqueness for [groups](https://github.com/satsuma-data/node-gateway/blob/de5533b84efa6b1f906479756f44f3393a9b96b6/internal/config/config.go#L122-L142), so I think we can do it for upstreamIDs (within a single chain config).  Maybe [here](https://github.com/satsuma-data/node-gateway/blob/de5533b84efa6b1f906479756f44f3393a9b96b6/internal/config/config.go#L162)? WDYT

Comment (kevin-satsuma) [reply internal/metadata/store.go:55]: Thanks for the links!  I ended up changing `GetUpstreamStatus` to use `groupID+upstreamID` as well.  Also added in validation in the places you linked for checking that upstream IDs are unique within a single group.

Comment (brianluong) [reply internal/metadata/store.go:55]: My bad, I think I was unclear - do we need to make the key "group ID + upstream ID"? Wouldn't upstreamID be enough?

Each _chain_ has its [own set of components](https://github.com/satsuma-data/node-gateway/blob/e0564fadc5bfde876c9b53977af40e94e2728290/internal/server/object_graph.go#L29)  (healthcheck manager, metadata store, etc). 

After the validation you added at the `config.go` level , there shouldn't be any overlap of upstream IDs.

**Review by brianluong — APPROVED** [review]

**Pull request merged**